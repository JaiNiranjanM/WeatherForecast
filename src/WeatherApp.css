.app {
  display: flex;
  width: 95%;
  flex-direction: column;
  margin: 15px auto;
  font-family: "Gill Sans", "Gill Sans MT", Calibri, "Trebuchet MS", sans-serif;
  color: #3e3e3e;
}

.today-weather-container {
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  height: 150px;
  background: rgb(2, 0, 36);
  background: linear-gradient(
    90deg,
    rgba(2, 0, 36, 1) 0%,
    rgba(46, 122, 65, 1) 0%,
    rgba(0, 212, 255, 1) 100%
  );
  border-radius: 5px;
}

.weather-image {
  display: flex;
  flex-direction: column;
}

.weather-image div {
  align-self: center;
  position: absolute;
  font-family: monospace;
  font-size: 14px;
}

.nextdays-weather-container {
  display: flex;
  width: 100%;
  margin-top: 20px;
  flex-direction: column;
  background: rgb(34, 193, 195);
  background: linear-gradient(
    0deg,
    rgba(34, 193, 195, 1) 0%,
    rgba(253, 201, 45, 1) 100%
  );
  border-radius: 20px;
}

.left-container {
  display: flex;
  align-items: center;
}

.main-temp {
  font-size: 70px;
  display: flex;
}

.main-temp-degree {
  margin-bottom: 20px;
}

.main-temp-loc {
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  margin-bottom: 20px;
}

.main-temp-loc-status {
  align-self: center;
  font-family: cursive;
  font-style: italic;
  opacity: 0.6;
}

.main-temp-measure {
  margin-bottom: 46px;
  margin-left: 10px;
  font-size: 16px;
  align-self: center;
}

.main-temp-extras {
  display: flex;
  flex-direction: column;
  margin-left: 15px;
}

.right-container {
  flex-direction: column;
  align-self: center;
  display: flex;
  margin-left: auto;
}

.right-container div {
  align-self: flex-end;
  margin-right: 40px;
}

.details-data {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 5px 15px;
}

.date-details {
  margin: 15px;
  background: rgb(2, 0, 36);
  background: linear-gradient(
    90deg,
    rgba(2, 0, 36, 1) 0%,
    rgba(46, 122, 65, 1) 0%,
    rgba(0, 212, 255, 1) 100%
  );
  width: 100px;
  text-align: center;
  border-radius: 10px;
  height: 26px;
  line-height: 26px;
}

.weather-daily-details {
  display: flex;
  flex-wrap: wrap;
  
}

@media screen and (max-width: 480px) {
  .today-weather-container {
    height: 180px;
    flex-wrap: wrap;
    padding: 15px 0;
  }

  .main-temp-extras {
    margin: 0;
    padding: 20px;
  }

  .main-temp-loc {
    align-items: flex-start;
  }
  
  .weather-daily-details {
    justify-content: space-around;
  }
}
