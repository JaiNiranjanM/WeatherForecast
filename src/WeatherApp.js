import React, { useState, useEffect } from "react";
import WeatherData from "./mockData";
import CountryCode from "./CountryCode";

import Forecast from "./Forecast";
import "./WeatherApp.css";

function WeatherApp() {
  const [currentWeatherData, setCurrentWeatherData] = useState({});
  const [countryCode, setCountryCode] = useState([]);
  const [weatherForecastData, setWeatherForecastData] = useState([]);
  const [availableDates, setAvailableDates] = useState([]);
  const [error, setError] = useState(null);
  const WEATHER_API_KEY = "524901&appid=532d5fa2a06b1cc30182238e6835eebb";
  const LOCATION = "bengaluru,karnataka,560093";

  useEffect(() => {
    const options = {
      headers: {
        "Referrer-Policy": "origin-when-cross-origin",
        "Access-Control-Allow-Origin": "https://api.openweathermap.org",
        "Access-Control-Allow-Methods": "POST",
        "Access-Control-Allow-Headers": "Content-Type, Authorization",
      },
      mode: "no-cors",
    };
    fetch(
      //api.openweathermap.org/data/2.5/weather?units=metric&q=" +
        LOCATION +
        WEATHER_API_KEY,
      options
    )
      .then((res) => res.json())
      .then(
        (result) => {
          setCurrentWeatherData(result);
          if (result && result.cod === ("400" || "500" || "501")) {
            setCurrentWeatherData(WeatherData.weatherData);
          }
        },
        (error) => {
          setError(error);
          setCurrentWeatherData(WeatherData.weatherData);
        }
      )
      .catch(() => setCurrentWeatherData(WeatherData.weatherData));
    fetch(
      "//api.openweathermap.org/data/2.5/forecast?units=metric&" +
        LOCATION +
        "&id=" +
        WEATHER_API_KEY,
      options
    )
      .then((res) => res.json())
      .then(
        (result) => {
          setWeatherForecastData(result);
          getForcastData(result);
          if (result && result.cod === ("400" || "500" || "501")) {
            setWeatherForecastData(WeatherData.forecastData);
            getForcastData(WeatherData && WeatherData.forecastData);
          }
        },
        (error) => {
          setError(error);
          setWeatherForecastData(WeatherData.forecastData);
          getForcastData(WeatherData && WeatherData.forecastData);
        }
      );

    setCountryCode(CountryCode);
  }, []);

  const getForcastData = (weatherForecastDetails) => {
    const availableDates =
      weatherForecastDetails &&
      weatherForecastDetails.list &&
      weatherForecastDetails.list
        .map((item) => item.dt_txt.split(" ")[0])
        .filter((value, index, self) => self.indexOf(value) === index);
    setAvailableDates(availableDates);
  };

  const getWeatherStatus = () => {
    const status = {
      description:
        currentWeatherData &&
        currentWeatherData.weather &&
        currentWeatherData.weather[0] &&
        currentWeatherData.weather[0].description,
      name:
        currentWeatherData &&
        currentWeatherData.weather &&
        currentWeatherData.weather[0] &&
        currentWeatherData.weather[0].main,
    };
    return status;
  };

  const getLogoIcon = () => {
    return (
      currentWeatherData &&
      currentWeatherData.weather &&
      currentWeatherData.weather[0] &&
      currentWeatherData.weather[0].icon
    );
  };

  const getWeatherUnit = () => {
    return (
      currentWeatherData &&
      currentWeatherData.main &&
      currentWeatherData.main.temp &&
      currentWeatherData.main.temp.toFixed(1)
    );
  };

  const getHumidity = () => {
    return (
      currentWeatherData &&
      currentWeatherData.main &&
      currentWeatherData.main.humidity
    );
  };

  const getPresure = () => {
    return (
      currentWeatherData &&
      currentWeatherData.main &&
      currentWeatherData.main.pressure
    );
  };

  const getWind = () => {
    return (
      currentWeatherData &&
      currentWeatherData.wind &&
      currentWeatherData.wind.speed
    );
  };

  const getLocation = () => {
    const country =
      currentWeatherData &&
      currentWeatherData.sys &&
      currentWeatherData.sys.country
        ? currentWeatherData.sys.country.toLowerCase()
        : "";
    const countryCodeData =
      countryCode && countryCode.filter((cntry) => cntry.code === country);
    const location = {
      city: currentWeatherData && currentWeatherData.name,
      countryCode: country,
      countryName:
        countryCodeData && countryCodeData[0] && countryCodeData[0].name,
    };

    return location;
  };

  return (
    <div className="app">
      {error ? (
        <div>API ERROR - Fetching the data from STATIC request</div>
      ) : (
        <div></div>
      )}
      <div className="today-weather-container">
        <div className="weather-image">
          <img
            src={
              getLogoIcon()
                ? `http://openweathermap.org/img/wn/${getLogoIcon()}@2x.png`
                : ""
            }
            alt="status-logo"
          />
          <div>{getWeatherStatus().description}</div>
        </div>

        <div className="main-temp">
          <div className="main-temp-degree">{getWeatherUnit()}</div>
          <div className="main-temp-measure">Â°C</div>
        </div>
        <div className="main-temp-extras">
          <div>Humidity: {getHumidity()}%</div>
          <div>Pressure: {getPresure()} Pa</div>
          <div>Wind: {getWind()} mph</div>
        </div>
        <div className="main-temp-loc">
          <div>
            {getLocation().city}, {getLocation().countryName}
          </div>
          <div className="main-temp-loc-status">{getWeatherStatus().name}</div>
        </div>
      </div>

      <div className="nextdays-weather-container">
        {availableDates &&
          availableDates.map((data, index) => {
            return (
              <Forecast
                availableDates={data}
                forecastData={weatherForecastData}
                index={index}
              />
            );
          })}
      </div>
    </div>
  );
}

export default WeatherApp;
