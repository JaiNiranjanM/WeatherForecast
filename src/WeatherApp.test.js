import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import WeatherApp from "./WeatherApp";
import WeatherData from "./mockData";

configure({ adapter: new Adapter() });

describe("WeatherApp Component", () => {
  let props;
  let wrapper;
  const currentWeatherData = WeatherData.weatherData;
  const weatherForecastData = WeatherData.forecastData;
  beforeEach(() => {
    const spy = jest.spyOn(React, "useEffect").mockImplementation((f) => f());

    props = {
      fetchCurrentWeatherData: jest.fn().mockResolvedValue(currentWeatherData),
      fetchWeatherForecastData: jest
        .fn()
        .mockResolvedValue(weatherForecastData),
    };

    wrapper = shallow(<WeatherApp {...props} />);
  });

  it("should render WeatherApp component", () => {
    const component = shallow(<WeatherApp />);
    expect(component.getElements()).toMatchSnapshot();
  });

  it("loads the CurrentWeatherData", () => {
    expect(props.fetchCurrentWeatherData).toBeCalledTimes(0);
  });

  it("loads the WeatherForecastData", () => {
    expect(props.fetchWeatherForecastData).toBeCalledTimes(0);
  });
});
