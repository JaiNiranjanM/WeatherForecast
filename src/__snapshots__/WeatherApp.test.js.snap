// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherApp Component should render WeatherApp component 1`] = `
Array [
  <div
    className="app"
  >
    <div />
    <div
      className="today-weather-container"
    >
      <div
        className="weather-image"
      >
        <img
          alt="status-logo"
          src=""
        />
        <div />
      </div>
      <div
        className="main-temp"
      >
        <div
          className="main-temp-degree"
        />
        <div
          className="main-temp-measure"
        >
          Â°C
        </div>
      </div>
      <div
        className="main-temp-extras"
      >
        <div>
          Humidity: 
          %
        </div>
        <div>
          Pressure: 
           Pa
        </div>
        <div>
          Wind: 
           mph
        </div>
      </div>
      <div
        className="main-temp-loc"
      >
        <div>
          , 
        </div>
        <div
          className="main-temp-loc-status"
        />
      </div>
    </div>
    <div
      className="nextdays-weather-container"
    />
  </div>,
]
`;
